### PR TITLE
expression: rewrite builtin function: CASE

### PR DESCRIPTION
**Pull request by spongedu**
for #4080

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
CI failed, please take a look.

**Comment by spongedu** [comment]
@XuHuaiyu   It seems that ci-tests compile fail due to  memory issues. I don't meet this problem on my machine. The same with PR #4164.

File: expression/builtin_control.go [hunk 1]
```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
...
+	// Fill in each 'THEN' clause parameter type.
+	fieldTps := make([]*types.FieldType, 0, (l+1)/2)
+	decimal, flen, isBinaryStr := args[1].GetType().Decimal, 0, false
+	for i := 1; i < l; i += 2 {
+		fieldTps = append(fieldTps, args[i].GetType())
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[i].GetType().Decimal), mathutil.Max(flen, args[i].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[i].GetType())
+	}
+	if l%2 == 1 {
+		fieldTps = append(fieldTps, args[l-1].GetType())
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[l-1].GetType().Decimal), mathutil.Max(flen, args[l-1].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[l-1].GetType())
```
Comment (zz-jason) [inline expression/builtin_control.go]: ditto

File: expression/builtin_control.go [hunk 2]
```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
...
+	fieldTp := types.AggFieldType(fieldTps)
+	classType := types.AggTypeClass(fieldTps, &fieldTp.Flag)
+	// Set retType to BINARY(0) if all arguments are of type NULL
+	if fieldTp.Tp == mysql.TypeNull {
+		fieldTp.Flen, fieldTp.Decimal = 0, 0
+		types.SetBinChsClnFlag(fieldTp)
+	}
+	fieldTp.Decimal, fieldTp.Flen = decimal, flen
+	var tp evalTp
+	switch classType {
```
Comment (zz-jason) [inline expression/builtin_control.go]: how about using this function to determine evalTp: `fieldTp2EvalTp`

File: expression/builtin_control.go [hunk 3]
```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
...
+	var (
+		arg    int64
+		isNull bool
+		err    error
+	)
+	args, l := b.getArgs(), len(b.getArgs())
+	for i := 0; i < l-1; i += 2 {
+		arg, isNull, err = args[i].EvalInt(row, sc)
+		if err != nil {
+			return 0, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_control.go]: when error being detected, why the returned value of isNull is `false` ?

File: expression/builtin_control.go [hunk 4]
```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
...
+			}
+			return arg, isNull, nil
+		}
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clause has else clause, l%2 == 1.
+	if l%2 == 1 {
+		arg, isNull, err = args[l-1].EvalInt(row, sc)
+		if err != nil {
```
Comment (zz-jason) [inline expression/builtin_control.go]: `return arg, isNull, errors.Trace(err)`

File: expression/builtin_control.go (hunk 3 above)
Comment (spongedu) [inline expression/builtin_control.go]: I think it doesn't matter whether `isNull == true` or `isNull == false` when `err != nil` here

File: expression/builtin_control.go (hunk 2 above)
Comment (spongedu) [inline expression/builtin_control.go]: There's mall differences between `fieldTp2EvalTp` and the `switch - case` here. `fieldTp.Decimal` needs to be set to 0 if `classType == ClassInt` and  `fieldTp.Charset, fieldTp.Collate ` need to be reset under certain condition

File: expression/builtin_control.go (hunk 4 above)
Comment (spongedu) [inline expression/builtin_control.go]: ok

File: expression/builtin_control.go (hunk 1 above)
Comment (spongedu) [inline expression/builtin_control.go]: ok

File: expression/builtin_control.go [hunk 5]
```diff
@@ -50,47 +55,345 @@ type caseWhenFunctionClass struct {
...
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinCaseWhenSig{newBaseBuiltinFunc(args, ctx)}
+	l := len(args)
+	// Fill in each 'THEN' clause parameter type.
+	fieldTps := make([]*types.FieldType, 0, (l+1)/2)
+	decimal, flen, isBinaryStr := args[1].GetType().Decimal, 0, false
+	for i := 1; i < l; i += 2 {
+		fieldTps = append(fieldTps, args[i].GetType())
+		decimal, flen, isBinaryStr = mathutil.Max(decimal, args[i].GetType().Decimal), mathutil.Max(flen, args[i].GetType().Flen), isBinaryStr || types.IsBinaryStr(args[i].GetType())
```
Comment (spongedu) [inline expression/builtin_control.go]: ok

File: expression/builtin_control.go (hunk 4 above)
Comment (zz-jason) [inline expression/builtin_control.go]: @spongedu any update ?

**Comment by jackysp** [comment]
Need to support json type? @hicqu

**Comment by jackysp** [comment]
Add some integration test.

File: expression/builtin_control.go [hunk 6]
```diff
@@ -50,47 +55,349 @@ type caseWhenFunctionClass struct {
...
+	}
+	if l%2 == 1 {
+		fieldTps = append(fieldTps, args[l-1].GetType())
+		decimal = mathutil.Max(decimal, args[l-1].GetType().Decimal)
+		flen = mathutil.Max(flen, args[l-1].GetType().Flen)
+		isBinaryStr = isBinaryStr || types.IsBinaryStr(args[l-1].GetType())
+	}
+	fieldTp := types.AggFieldType(fieldTps)
+	classType := types.AggTypeClass(fieldTps, &fieldTp.Flag)
+	// Set retType to BINARY(0) if all arguments are of type NULL
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: add `.` at the end.

File: expression/builtin_control.go [hunk 7]
```diff
@@ -50,47 +55,349 @@ type caseWhenFunctionClass struct {
...
+		argTps = append(argTps, tpInt, tp)
+	}
+	if l%2 == 1 {
+		argTps = append(argTps, tp)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tp, argTps...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: remove this.

File: expression/builtin_control.go [hunk 8]
```diff
@@ -50,47 +55,349 @@ type caseWhenFunctionClass struct {
...
+		arg, isNull, err = args[i].EvalInt(row, sc)
+		if err != nil {
+			return 0, isNull, errors.Trace(err)
+		}
+		if isNull {
+			continue
+		}
+		if arg != 0 {
+			arg, isNull, err = args[i+1].EvalInt(row, sc)
+			if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: 1.we do not need to check this
`return arg, isNull, errors.Trace(err) ` is enough

2. so as the others

File: expression/builtin_control.go [hunk 9]
```diff
@@ -50,47 +55,349 @@ type caseWhenFunctionClass struct {
...
+func (b *builtinCaseWhenIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var (
+		arg    int64
+		isNull bool
+		err    error
+	)
+	args, l := b.getArgs(), len(b.getArgs())
+	for i := 0; i < l-1; i += 2 {
+		arg, isNull, err = args[i].EvalInt(row, sc)
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: s/ arg/ condition

File: expression/builtin_control.go [hunk 10]
```diff
@@ -50,47 +55,349 @@ type caseWhenFunctionClass struct {
...
+		if err != nil {
+			return 0, isNull, errors.Trace(err)
+		}
+		if isNull {
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]:
``` go
if isNull || arg == 0{
    continue
}
```
so line 162 can be removed.
so as the others.

File: expression/builtin_control.go (hunk 6 above)
Comment (spongedu) [inline expression/builtin_control.go]: ok

File: expression/builtin_control.go (hunk 7 above)
Comment (spongedu) [inline expression/builtin_control.go]: ok

File: expression/builtin_control.go (hunk 9 above)
Comment (spongedu) [inline expression/builtin_control.go]: WHEN `condition` and THNE `statement` share the same variable `arg` here, so I think it's not proper to rename `arg` as condition. We can use two variables `condition` and `statement` to represent them, or maintain the current form. I prefer the latter.

File: expression/builtin_control.go (hunk 10 above)
Comment (spongedu) [inline expression/builtin_control.go]: done

File: expression/builtin_control.go (hunk 8 above)
Comment (spongedu) [inline expression/builtin_control.go]: done

File: expression/builtin_control.go (hunk 9 above)
Comment (XuHuaiyu) [inline expression/builtin_control.go]: I think that keep the variable name be consistent with the others will be better ?

Comment (spongedu) [inline expression/builtin_control.go]: ok.

File: expression/builtin_control.go [hunk 11]
```diff
@@ -48,47 +53,312 @@ type caseWhenFunctionClass struct {
...
+// evalInt evals a builtinCaseWhenIntSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/case.html
-func (b *builtinCaseWhenSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinCaseWhenIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var (
+		condition, ret int64
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: move `ret, isNull, and err` to the return statement?
`func (b *builtinCaseWhenIntSig) evalInt(row []types.Datum) (ret int64, isNull bool, err error) `

File: expression/builtin_control.go [hunk 12]
```diff
@@ -48,47 +53,312 @@ type caseWhenFunctionClass struct {
...
-			return d, errors.Trace(err1)
+		ret, isNull, err = args[i+1].EvalDecimal(row, sc)
+		return ret, isNull, errors.Trace(err)
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clause has else clause, l%2 == 1.
+	if l%2 == 1 {
+		ret, isNull, err = args[l-1].EvalDecimal(row, sc)
+		if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: line 239 ~ line 242 is useless?

just use `return` in line 244 is enough?

so as the others.

File: expression/builtin_control.go (hunk 11 above)
Comment (spongedu) [inline expression/builtin_control.go]: ok

File: expression/builtin_control.go (hunk 12 above)
Comment (spongedu) [inline expression/builtin_control.go]: done

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL.

File: expression/builtin_control.go [hunk 13]
```diff
@@ -48,47 +53,268 @@ type caseWhenFunctionClass struct {
...
+		return ret, isNull, errors.Trace(err)
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clause has else clause, l%2 == 1.
+	if l%2 == 1 {
+		ret, isNull, err = args[l-1].EvalString(row, sc)
+		return ret, isNull, errors.Trace(err)
+	}
+	return "", true, nil
```
Comment (XuHuaiyu) [inline expression/builtin_control.go]: return ret, true, nil
so as the others

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**